crates/linker/src/generate_dylib/pe.rs

```diff
@@ -0,0 +1,160 @@
+        minor_version: U16::new(LE, 0),
+        name: U32::new(LE, virtual_address + directory_size),
+        base: U32::new(LE, 0),
+        number_of_functions: U32::new(LE, custom_names.len() as u32),
+        number_of_names: U32::new(LE, custom_names.len() as u32),
+        address_of_functions: U32::new(LE, address_of_functions),
+        address_of_names: U32::new(LE, address_of_names),
+        address_of_name_ordinals: U32::new(LE, address_of_name_ordinals),
+    }
+}
+
+fn synthetic_export_dir(virtual_address: u32, custom_names: &[String]) -> Vec<u8> {
+    let mut vec = vec![0; std::mem::size_of::<pe::ImageExportDirectory>()];
+
+    let ptr = vec.as_mut_ptr();
+
+    let name = "roc-cheaty-lib.dll";
+    let directory = synthetic_image_export_directory(name, virtual_address, custom_names);
+
+    unsafe {
+        std::ptr::write_unaligned(ptr as *mut pe::ImageExportDirectory, directory);
+    }
+
+    // write the .dll name, null-terminated
+    vec.extend(name.as_bytes());
+    vec.push(0);
+
+    let n = custom_names.len();
+
+    // Unsure what this one does; it does not seem important for our purposes
+    //
+    // Export Address Table -- Ordinal Base 0
+    //  [   1] +base[   1] 1020 Export RVA
+    //  [   2] +base[   2] d030 Export RVA
+    for _ in custom_names {
+        vec.extend(42u32.to_le_bytes());
+    }
+
+    // Maps the index to a name
+    //
+    // [Ordinal/Name Pointer] Table
+    //  [   1] _CRT_INIT
+    //  [   2] _CRT_MT
+    let mut acc = directory.address_of_name_ordinals.get(LE) + n as u32 * 2;
+    for name in custom_names {
+        vec.extend(acc.to_le_bytes());
+        acc += name.len() as u32 + 1;
+    }
+
+    // the ordinals, which just map to the index in our case
+    for (i, _) in custom_names.iter().enumerate() {
+        vec.extend((i as u16).to_le_bytes());
+    }
+
+    // write out the names of the symbols, as null-terminated strings
+    for name in custom_names {
+        vec.extend(name.as_bytes());
+        vec.push(0);
+    }
+
+    vec
+}
+
+pub fn synthetic_dll(custom_names: &[String]) -> Vec<u8> {
+    let mut out_data = Vec::new();
+    let mut writer = object::write::pe::Writer::new(true, 8, 8, &mut out_data);
+
+    // fairly randomly chosen. Not sure if this is relevant
+    let virtual_address = 0x138;
+
+    let exports = synthetic_export_dir(virtual_address, custom_names);
+
+    // Reserve file ranges and virtual addresses.
+    writer.reserve_dos_header_and_stub();
+
+    // we will have one header: the export directory
+    writer.reserve_nt_headers(1);
+
+    writer.set_data_directory(
+        pe::IMAGE_DIRECTORY_ENTRY_EXPORT,
+        virtual_address,
+        exports.len() as _,
+    );
+
+    writer.reserve_section_headers(1);
+
+    // we store the export directory in a .rdata section
+    let rdata_section: (_, Vec<u8>) = {
+        let range = writer.reserve_section(
+            *b".rdata\0\0",
+            1073741888,
```

Comment:
Is this always accurate?

Reply:
I'm assuming you mean the random number there? I think it's always valid, it sets a bunch of bitflags I guess
```
    pub fn reserve_section(
        &mut self,
        name: [u8; 8],
        characteristics: u32,
        virtual_size: u32,
        data_size: u32,
```